nifi-nar-bundles/nifi-provenance-repository-bundle/nifi-persistent-provenance-repository/src/main/java/org/apache/nifi/provenance/serialization/CompressableRecordReader.java

```diff
@@ -0,0 +1,277 @@
+            if (offset < 0) {
+                limitedStream = rawInputStream;
+            } else {
+                limitedStream = new LimitingInputStream(rawInputStream, offset - rawInputStream.getBytesConsumed());
+            }
+        }
+
+        final InputStream readableStream;
+        if (compressed) {
+            readableStream = new BufferedInputStream(new GZIPInputStream(limitedStream));
+        } else {
+            readableStream = new BufferedInputStream(limitedStream);
+        }
+
+        byteCountingIn = new ByteCountingInputStream(readableStream, rawInputStream.getBytesConsumed());
+        dis = new DataInputStream(byteCountingIn);
+    }
+
+
+    @Override
+    public TocReader getTocReader() {
+        return tocReader;
+    }
+
+    @Override
+    public boolean isBlockIndexAvailable() {
+        return tocReader != null;
+    }
+
+    @Override
+    public int getBlockIndex() {
+        if (tocReader == null) {
+            throw new IllegalStateException("Cannot determine Block Index because no Table-of-Contents could be found for Provenance Log " + filename);
+        }
+
+        return tocReader.getBlockIndex(rawInputStream.getBytesConsumed());
+    }
+
+    @Override
+    public long getBytesConsumed() {
+        return byteCountingIn.getBytesConsumed();
+    }
+
+    private boolean isData() throws IOException {
+        byteCountingIn.mark(1);
+        int nextByte = byteCountingIn.read();
+        byteCountingIn.reset();
+
+        if (nextByte < 0) {
+            try {
+                resetStreamForNextBlock();
+            } catch (final EOFException eof) {
+                return false;
+            }
+
+            byteCountingIn.mark(1);
+            nextByte = byteCountingIn.read();
+            byteCountingIn.reset();
+        }
+
+        return nextByte >= 0;
+    }
+
+    @Override
+    public long getMaxEventId() throws IOException {
+        if (tocReader != null) {
+            final long lastBlockOffset = tocReader.getLastBlockOffset();
+            skipToBlock(tocReader.getBlockIndex(lastBlockOffset));
+        }
+
+        ProvenanceEventRecord record;
+        ProvenanceEventRecord lastRecord = null;
+        try {
+            while ((record = nextRecord()) != null) {
+                lastRecord = record;
+            }
+        } catch (final EOFException eof) {
+            // This can happen if we stop NIFi while the record is being written.
+            // This is OK, we just ignore this record. The session will not have been
+            // committed, so we can just process the FlowFile again.
+        }
+
+        return lastRecord == null ? -1L : lastRecord.getEventId();
+    }
+
+    @Override
+    public void close() throws IOException {
+        logger.trace("Closing Record Reader for {}", filename);
+
+        dis.close();
+        rawInputStream.close();
+
+        if (tocReader != null) {
+            tocReader.close();
+        }
+    }
```

Comment:
Should all _close()_ calls be wrapped in try/catch. Worried about resource leaks if any (other then last) fail.